src/qibo/gates/abstract.py

```diff
@@ -204,7 +204,8 @@ def _find_repeated(qubits: Sequence[int]) -> int:
     def _check_control_target_overlap(self):
         """Checks that there are no qubits that are both target and
         controls."""
-        common = set(self._target_qubits) & self._control_qubits
+        control_target = self._target_qubits + self._control_qubits
+        common = len(set(control_target)) != len(control_target)
         if common:
             raise_error(
                 ValueError,
```

Comment:
> Other than that, what is the motivation of switching `_control_qubits` from `set` to `tuple`? Just symmetry with `_target_qubits`? (not intending to say that the current code is the best, I am sure it can be simplified in many ways)

```python

In [1]: %timeit set()
55.7 ns ± 1.21 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [2]: %timeit tuple()
33.1 ns ± 0.247 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [3]: %timeit ()
8.74 ns ± 0.179 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)
```
It's 7x faster, even though it's in the `ns` scale, it makes a difference when creating multiple thousands of gates e.g. `random_clifford` for 1000+ qubits.

Reply:
Another thing that we noted is that every single qubit gate is inheriting a `control_qubits` attribute from the abstract `Gate` class. This means that every single qubit gate carries an empty tuple with it, unless the method `controlled_by` is used and the tuple is filled. It may be worth for single qubit gates to move the initialization of `control_qubits` to inside the `control_qubits` method. That would make the creating of gates slightly faster and the objects slightly lighter.